### DISCUSSION TITLE
OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE - Macro usage

### DISCUSSION THREAD

**Discussion by jchellap** [openthread/openthread#9323] [General]
Hi All,

OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE - should we enable this macro only for NCP platforms or applies to RCP platforms also? On the platform side, what should be the corresponding implementation(any pointers will help) to be done?
Is there any example code which can be looked at the usage point of view?
Could someone elaborate on this to get more insight on the usage of the same?

**Comment by Irving-cl** [comment]
Hi @jchellap,

> should we enable this macro only for NCP platforms or applies to RCP platforms also?

It doesn't depend on NCP or RCP. It depends on if your board (embedded system) has coex radio. For example, if wifi is also located on the board, you may want to handle the coex case. However the implementation of coex handling is totally customized. It is out of the scope of OT and should be provided by the manufacturer. For example, a new product is designed: there are both Wifi SoC and Thread SoC on its board, the two SoCs are connected using 3-wires (REQ, PRI, GRANT) which is a common coex solution. Then the manufacturer should add their implementation in their Thread firmwares and Wifi firmwares.

So in all examples provided by OT, you don't need to enable this macro.

The platform coex functions defined by OT are:
* Turn on/off coex provided by the platform. If turning on, the Thread tx/rx should be coordinated by the coex solution provided by the platform. If turning off, the Thread tx/rx will do as is. But if your device has other co-located 2.4Ghz radio, you might have coex issues. (For example, wifi Tx is going while Thread Tx is also going. The signals will impact each other)
* Fetch coex metrics. OT specified a set of coex metrics. The platform should implement the function to count and return these metrics.

> what should be the corresponding implementation(any pointers will help) to be done

I assume you mean the platform coex function(**not** the OT platform coex functions: turn on/off and provide coex metrics). As I mentioned, it's totally dependent on the product and hardware design.

Let's think of an 3-wire example with a Thread SoC and a Wifi SoC on the board and the two SoCs are connected with 3 wires (REQ,PRI and GRANT). Then you (as vendor of Thread SoC) need to implement things like: before doing Thread TX, checking if GRANT is asserted. If GRANT is asserted, do Thread TX. If not, based on your algorithm, wait for the GRANT to be asserted or assert PRI to add priority, etc. The coex logic is what you need to implement.

Of course, we will also need to implement OT platform coex function: turn on/off coex and provide coex metrics. You can see the implementation on [simulation platform](https://github.com/openthread/openthread/blob/main/examples/platforms/simulation/radio.c#L1236).

**Reply by jchellap** [reply]
Thanks @Irving-cl 
Yes we have a combo chip and the coex functionality level at the platform is implemented. I was looking more at the plugging in Open Thread Coex solution with the platform coex implementation.
Thanks for the clear description of how it works.
So, i understand that at the platform level we need to turn on coex (using OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE)  and provide Coex metrics to get the coex working.
Let me look at Simulation platform... thanks for the guidance.

**Reply by jchellap** [reply]
on Another note, just to clarify one more point, we have a combo chip with Coex implemented at platform level and if we don't enable the macro OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE, do we miss anything in CoEx working or if the SoC is combo, open thread macro OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE should be enabled with the necessary API implemented to coexist?

**Reply by Irving-cl** [reply]
> if we don't enable the macro OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE, do we miss anything in CoEx working or if the SoC is combo, open thread macro OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE should be enabled with the necessary API implemented to coexist?

We recommend to implement the OT platform coex API. But it's not a MUST. It depends on your need. If `OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE` isn't enabled, the impact is that we cannot turn on/off the coex from OT and we cannot fetch the coex metrics. It doesn't matter if the intention is just to always keep coex function working on the FW. So even `OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE` is disabled, the coex function implemented by the platform will still work. 

From my experience, we would sometimes test and compare the cases where coex is enabled and disabled and analyze the metrics. So we usually implement the OT platform coex APIs and enable this option. @jchellap

**Reply by jchellap** [reply]
@Irving-cl , Thanks!!
